agg-ast/schema_derivation/src/schema_derivation.rs

```diff
@@ -35,21 +37,203 @@ pub(crate) struct ResultSetState<'a> {
     pub null_behavior: Satisfaction,
 }
 
+fn derive_schema_for_pipeline(pipeline: Vec<Stage>, state: &mut ResultSetState) -> Result<Schema> {
```

Comment:
this is a generic helper that I believe we will also be able to use for the overall api in milestone 4!